src/book/02-system/01-foundations/07-director.mdx

```diff
@@ -0,0 +1,231 @@
+
+Example:
+
+```cpp
+on<Provides<Walk>>().then()
+```
+
+#### Leaving
+
+Leaving providers are used for the state when a provider group is losing control and will no longer have access to subtasks, or when a higher priority task elsewhere needs a causing that it provides.
+Leaving providers are executed in two scenarios:
+
+Firstly, A leaving provider will be executed once when the provider no longer has an active task.
+Note: If there no leaving providers that are satisfied in this situation (due to when conditions) the system will stop executing the provider without informing it.
+It may be advantageous to have a `Leaving` provider with no when conditions if you want to always know when you no longer have a task.
+
+Secondly, a leaving provider may be executed as a result of it having a causing reaction that another task needs if it has a higher priority than this task.
+For example, if a higher priority kick engine wanted to kick being staticallys stable, and the walk engine had a leaving which caused statically stable, then the leaving .
+In the event that the kick providers task had a lower priority than this providers task, then it is only able to change this module to run other non `Leaving` providers to achive its causing.
+
+In order for a leaving provider to be run however, it must have first been in the entering or normal states.
+The system cannot immediately jump to a leaving provider.
+
+Example:
+
+```cpp
+on<Provides<Walk>>().then()
+```
+
+### When
+
+A when condition is used to ensure that a provider will only execute when its conditions are met.
+These conditions are provided using a global enum message along with a statement that describes what condition you want.
+For example if you wanted a kick engine to only run if the robot is statically stable or better.
+
+```cpp
+on<Provides<Kick>, When<Stability, std::greater_equal, Stability::STATIC>>().then(...);
+```
+
+These when conditions are updated by using normal NUClear emits.
+
+### Causing
+
+A causing statement is a promise that by running this provider, the desired state will be eventually reached.
+This is used to allow smooth transitions beween modules.
+For example when the walk engine stopped moving it would be in a "Standing" stability state.
+This means the walk engine could have a `Leaving` reaction that caused this state so other modules know it can use this to be standing before they execute their actions.
+
+```cpp
+on<Leaving<Walk>, Causing<Stability, Stability::STANDING>().then(...);
+```
+
+Other providers will take advantage of this in two ways.
+If the other provider has higher priority than this one, it will force this provider group to switch to a `Provides` or `Leaving` that has a causing that satisfies the when condition it is limited on.
+If the other provider has lower priority than this one, it will only be able to force this provider to change to another `Provides` statement and not a leavign one.
```

Comment:
```suggestion
If the other provider has lower priority than this one, it will only be able to force this provider to change to another `Provides` statement and not a leaving one.
```

Reply:
Right so if a kick has a higher priority to the walk then it can force the walk to change to the leaving provider that causes standing/stability.

Reply:
Correct, that way the kick doesn't have to care about "how" it's getting to the state it needs. It just needs to know it's the most important thing to do right now and the director will try to find a way to make it work.

So that means no hacky things in the kick engine or another module where you monitor when the walk has finished walking so you can activate the kick etc, it'll just work itself out

Reply:
Sounds good.